pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			var found bool
+			if state.IsWildcard, found = isWildcard[req.TypeUrl]; !found {
+				state.IsWildcard = len(req.GetResourceNamesSubscribe()) == 0
+				isWildcard[req.TypeUrl] = state.IsWildcard
+			}
+
+			if sub := req.GetResourceNamesSubscribe(); len(sub) > 0 {
+				s.subscribe(sub, state.ResourceVersions)
+			}
+			for r, v := range req.InitialResourceVersions {
+				state.ResourceVersions[r] = v
+			}
+			if unsub := req.GetResourceNamesUnsubscribe(); len(unsub) > 0 {
+				s.unsubscribe(unsub, state.ResourceVersions)
+			}
+
+			if s.callbacks != nil {
+				if err := s.callbacks.OnStreamDeltaRequest(streamID, req); err != nil {
+					return err
+				}
+			}
+
+			// cancel existing watches to (re-)request a newer version
+			switch {
+			case req.TypeUrl == resource.EndpointType:
+				if values.deltaEndpointNonce == "" || values.deltaEndpointNonce == nonce {
+					if values.deltaEndpointCancel != nil {
+						values.deltaEndpointCancel()
+					}
+					values.deltaEndpoints, values.deltaEndpointCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.ClusterType:
+				if values.deltaClusterNonce == "" || values.deltaClusterNonce == nonce {
+					if values.deltaClusterCancel != nil {
+						values.deltaClusterCancel()
+					}
+					values.deltaClusters, values.deltaClusterCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RouteType:
+				if values.deltaRouteNonce == "" || values.deltaRouteNonce == nonce {
+					if values.deltaRouteCancel != nil {
+						values.deltaRouteCancel()
+					}
+					values.deltaRoutes, values.deltaRouteCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.ListenerType:
+				if values.deltaListenerNonce == "" || values.deltaListenerNonce == nonce {
+					if values.deltaListenerCancel != nil {
+						values.deltaListenerCancel()
+					}
+					values.deltaListeners, values.deltaListenerCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.SecretType:
+				if values.deltaSecretNonce == "" || values.deltaSecretNonce == nonce {
+					if values.deltaSecretCancel != nil {
+						values.deltaSecretCancel()
+					}
+					values.deltaSecrets, values.deltaSecretCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RuntimeType:
+				if values.deltaRuntimeNonce == "" || values.deltaRuntimeNonce == nonce {
+					if values.deltaRuntimeCancel != nil {
+						values.deltaRuntimeCancel()
+					}
+					values.deltaRuntimes, values.deltaRuntimeCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			default:
+				typeURL := req.TypeUrl
+				responseNonce, seen := values.deltaNonces[typeURL]
+				if !seen || responseNonce == nonce {
+					// We must signal goroutine termination to prevent a race between the cancel closing the watch
+					// and the producer closing the watch.
+					if terminate, exists := values.deltaTerminations[typeURL]; exists {
```

Comment:
Since this code matches the sotw can we somehow reuse it? Or do the types not line up right?

Reply:
Sadly the types don't line up right 😞 I tried to reuse this to keep the code cleaner but it's stuck there